[SofaKernel] issofa_topology: Improvement, BugFix and Cleaning on Topology

guparan opened the pull request:
### New features
- UncoupledConstraintCorrection : add option to disable handling of topological changes for compliance vector
- Container : add accessor to the topology type
- SubsetMapping : add resizeToModel flag in SubsetMapping to enable resizing of the destination MechanicalState to the size of indices
- BaseContext : add new method getActiveMeshTopology to get relevant topology, taking `BaseMaping::sameTopology()` into account
- PointSetTopology : add some unit tests

### Improvements
- each time a topological change sequence is detected in BarycentricMapperTriangleSetTopology the mapping is initialized
- PointSetTopologyContainer maintains a list of point indices, accessible through the data points
- Topology : Factorize the last method specific to each topological elements in data containers

### Cleans
- MechanicalObject size and link to topology is now exposed as Data and Link, using the new getActiveMeshTopology() method. Set the link manually and/or set useTopology="false" to override

### Bugfixes
- apply data container updates before mapping updates
- use existing parameter handleTopologyChange in SubsetMapping
- slightly cleanup topology change handling method for triangle mappers
- check of upper topology based on the emptiness of the upper container
- crash in `MeshTopology::init()` for non manifold meshes
- re-enable Propagate of topological changes inbetween triangle creation and removal in `TriangleSetTopologyModifier::addRemoveTriangles(),` as it is required in cases involving chains of topology mapping and topology data containers depending on each other 

This is the re-opening of [previous issofa_topology pull-request](https://github.com/sofa-framework/sofa/pull/217).


______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings nor unit test failures.
- [x] does not break existing scenes.
- [x] does not break API compatibility.
- [x] has been reviewed and agreed to be transitional.
- [x] is more than 1 week old.

**Reviewers will merge only if all these checks are true.**

guparan committed:
Reopen issofa_topology pull-request

Modifications were reverted in master by commit 0dbf1ab50e1f69e583dd0bcf4437c75071366dcc after a wrong merge of PR #217.

See PR #243

guparan commented:
[ci-build]

guparan committed:
[SofaCUDA] FIX missing MechanicalObject::vsize

MechanicalObject::vsize was changed to MechanicalObject::d_size
in 99dc2a43b5f4e931a6daf27cadc617222de9553b

guparan committed:
[SofaOpenCL] FIX missing MechanicalObject::vsize

MechanicalObject::vsize was changed to MechanicalObject::d_size
in 99dc2a43b5f4e931a6daf27cadc617222de9553b

matthieu-nesme commented on SofaKernel/framework/sofa/core/topology/BaseMeshTopology.h:
```diff
@@ -132,7 +132,9 @@ class SOFA_CORE_API BaseMeshTopology : public core::topology::Topology
-    virtual const Hexa getHexahedron(HexaID i)       { return getHexahedra()[i]; }
+    virtual const Hexa getHexahedron(HexaID i)       { return getHexahedra()[i]; }   
+	   
+    virtual sofa::core::topology::TopologyObjectType getTopologyType() const = 0 ;
```
doc?

matthieu-nesme replied:
Still requires a description.
For example it is unclear to me what does this function return when the topology contains both triangles and quads or tetra and hexa?
But I do no see the definition of TopologyObjectType in this PR so maybe it is something already existing and already undocumented and it is fine like that....

JeremieA replied:
This method is not actually used, and I haven't checked why it was added exactly, but it looks like the main benefit is to get rid of codes that include and use chains of dynamic casts to *SetTopologyContainer to discover which type of topology it is, and to get expose the "UpperTopology" info already computed in MeshTopology that have a similar (undocumented) semantic. In case of *SetTopologyContainer, the semantic is clear, as you cannot mix triangles and quads. It is indeed unclear for MeshTopology, but that is nothing new (as this change is not computing it, just exposing it in a unified way).
I would rather have this method without comment and have a todo somewhere to clarify MeshTopology::UpperTopology, document it and this method, and add unit test(s), than to add an unclear comment here...

matthieu-nesme commented on SofaKernel/modules/SofaBaseMechanics/MechanicalObject.h:
```diff
@@ -422,8 +417,11 @@ protected :
-    sofa::core::objectmodel::DataFileName filename;
-    Data< bool> ignoreLoader;
+    //int vsize; ///< Number of elements to allocate in vectors
+    Data< int > d_size;
```
Why not but... why make it a Data?
What does happen if this Data value changes?
Note that a DataTracker could be used to automatically resize the mstate as soon as this Data is changed.

JeremieA replied:
We had one too many times where we had no idea if this hidden vsize value was correct, as well as the topology pointer.
The point of making them a proper Data<> and Link<> is to expose then for inspection.
Note that is Sofa we have no reliable way to protect a Data from external changes. The ReadOnly flag is only respected by the GUI, not from other component's code. But this issue is not specific to this Data.

matthieu-nesme replied:
ok.
Be aware of the potential conflict with `if (arg->getAttribute("size")` in `MechanicalObject::parse`.

matthieu-nesme replied:
Is the potential conflict handled?

JeremieA replied:
I don"t see the conflict. If a "size" is given to parse(), then it will be used to call resize(), as was the case before. The only addition is that the Base::parse() will parse it again and put the same value in the Data, but they are the same so it is harmless. The rest of the (broken) logic in init() is unaffected by exposing vsize as a Data instead of a hidden member.

matthieu-nesme replied:
Very good, thanks for the answer.

matthieu-nesme commented on SofaKernel/modules/SofaBaseMechanics/MechanicalObject.h:
```diff
@@ -475,7 +471,6 @@ protected :
     std::ofstream* m_gnuplotFileX;
     std::ofstream* m_gnuplotFileV;
 
-    sofa::core::topology::BaseMeshTopology* m_topology;
```
thank you ^^

matthieu-nesme commented on SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl:
```diff
@@ -103,23 +104,18 @@ MechanicalObject<DataTypes>::MechanicalObject()
     , m_gnuplotFileV(NULL)
 {
-    // HACK
-    if (!restScale.isSet())
```
what was the purpose of this hack?
It looks like it was just a way to always mark the Data as set, is it still relevant or not?

JeremieA replied:
The Data is constructed here, there is now way it can be set at this point. So this is same as if(true)...
I have no idea why it was here in the first place. A guess would be that some code somewhere was not working if restScale was not flagged as set, and this hack ensures that this never happen...

matthieu-nesme commented on SofaKernel/modules/SofaBaseMechanics/SubsetMapping.inl:
```diff
@@ -187,7 +191,15 @@ template <class TIn, class TOut>
+    
+    if (f_resizeToModel.getValue() || this->toModel->getSize() < indices.size())
+    { 
+        if (this->toModel->getSize() != indices.size()) 
```
You are comparing a != after a < (always true).

Personally, I would do like any other mappings, i.e. removing the f_resizeToModel Data (that should be named with a d_ BTW) and simply perform `this->toModel->resize(indices.size()); ` all the time. It would be way more clear, and possibly more efficient than evaluating Data and the user would not have to activate anything special.

JeremieA replied:
There are existing scenes that rely on SubsetMapping not resizing to the size of the indices array, when for instance creating 20 points but only the first 10 are updated by the mapping.

The design of most changes in Sofa in our branches is to never break existing run-time behavior of such commonly-used component, new behaviors (such as resizing here, or the disabling of the topo engine) are opt-in through an added flag).

If there is a decision regarding the behavior of standard mappings to resize or not the output model, then existing mapping should be updated but with a way to warn developers if they are potentially affected by the change of default behavior.

matthieu-nesme reviewed (changes requested):
I let someone-else reviewing the Topologies themselves.

epernod commented on SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl:
```diff
@@ -1087,12 +1085,15 @@ void MechanicalObject<DataTypes>::init()
 
+    if (l_topology)
+    {
+        sout << "Initialization with topology " << l_topology->getTypeName() << " " << l_topology->getName() << " ( " << l_topology->getNbPoints() << " points)" << sendl; // << (l_topology->hasPos() ? "WITH" : "WITHOUT") << " positions)"
```
Debug log to be removed? or maybe use new msg_ formating.

JeremieA replied:
The purpose was the have a developer-friendly message at the top of the component's log that would catch his attention if the implicitly-defined-with-tricky-rules association to a topology was incorrect. But it could indeed be removed.
We did not switch to the new msg_ formating due to conflicts with earlier changes on our branch and pending review of this quite complex new logger mechanism. But feel free to patch it if you know the right way to make it stay only visible if you look for it within the component's log and not the general console.

damienmarchal replied:
In the sofa-consortium master's branch sout is internally using msg_info(), in a similar way serr is in fact using msg_warning(). So there is absolutely no interest in using them and anyone can safely replace sout <<  by msg_info() and serr << msg_warning() because it will not change the behavior (and it brings several benefits).
 
Having the messages routed to the component's log and/or general console is not decided at the components level but done by the "MessagesHandler". Currently in runSofa the messages handlers installed by default are printing all the messages in the console and add them to the component's log (which is used by the GUI to 'nicely' display the messages in the panels).

epernod commented on SofaKernel/modules/SofaBaseMechanics/SubsetMapping.inl:
```diff
@@ -169,9 +170,12 @@ void SubsetMapping<TIn, TOut>::init()
+        // Initialize functions and parameters for topological changes
+        f_indices.createTopologicalEngine(topology);
+        f_indices.registerTopologicalData();
+    }
```
What happen if the Data<bool> f_resizeToModel is set to false, and the topology is still handling topologyChange? Will the list of indices of this subset still relevant ? Maybe you should add a condition on f_resizeToModel here.

JeremieA replied:
This is intended, the developer has the ability to set both flags on or off based on the scene design.
In most of our scenes it's not the responsability of the MechanicalMapping to resize the output model but rather it's the propagation of the topological changes up to each MechanicalState that is in charge of resizing it. The mapping should only handle its own internal data, except if someone else is already doing it (like if this list of indices is linked to the output of an engine such as BoxROI or from an associated TopologicalMapping).

epernod commented on SofaKernel/modules/SofaBaseTopology/PointSetTopologyContainer.cpp:
```diff
@@ -33,6 +36,25 @@ namespace component
+    PointID operator() () { return current++; }
+
+    PointID current;
+};
```
Just for curiosity, why using this structure instead of a simple int counter ? Is it for the algorithm to allocate consecutive memory?

JeremieA replied:
This is probably in order to be able to use std::generate() instead of a manual loop construct. Note that these changes were made before the decision to allow for c++11 constructs, it could potentially be replaced by a functor now.

epernod commented on SofaKernel/modules/SofaBaseTopology/TopologyData.inl:
```diff
@@ -57,7 +57,7 @@ void TopologyDataImpl <TopologyElementType, VecT>::createTopologicalEngine(sofa:
         if (this->getOwner() && dynamic_cast<sofa::core::objectmodel::BaseObject*>(this->getOwner())) dynamic_cast<sofa::core::objectmodel::BaseObject*>(this->getOwner())->addSlave(this->m_topologicalEngine.get());
         this->m_topologicalEngine->init();
-        this->linkToElementDataArray();
+        this->linkToElementDataArray((TopologyElementType*)NULL);
```
Small understanding question: According to the changes done, if here `this` is a `PointData`, passing a NULL Pointer will still call `linkToPointDataArray` ?

JeremieA replied:
The idea is to get rid of the need to have per-element-type child classes to TopologyDataImpl for the only purpose of overloading linkToElementDataArray(). Now instead this method takes a (null) pointer with the type of the element, such that the right one is called, based on the already-known template type. So yes it will call linkToPointDataArray, but only if TopologyElementType is indeed Point.

epernod reviewed (commented):
Except for the subsetMapping, those comments are more  to understand the changes.

damienmarchal commented:
Hi all, 

This PR is getting old, would it be possible to:
- Fix the conflict
- Update the reviewer's status if it is now ok

Thanks. 

EDIT: @matthieu-nesme Are you still requesting changes on this PR ?

guparan committed:
Merge remote-tracking branch 'origin/master' into issofa_topology

# Conflicts:
#	SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl
#	SofaKernel/modules/SofaBaseTopology/SofaBaseTopology_test/CMakeLists.txt

guparan committed:
[SofaBaseTopology] FIX wrong merge of CMakeLists

matthieu-nesme commented:
@damienmarchal I already added comments w/o answer, but I commented them again

JeremieA commented:
Answered (and please remember that all this was already reviewed and discussed during STC, the remaining issue was supposed to be only to fix sofaCUDA as far as I understood)

damienmarchal commented:
@matthieu-nesme and @JeremieA many thanks for the quick updates. 

@guparan I see you made a commit (15 May) about SofaCUDA. Does it means this pending issue is solved and we can merge or is there still something to do ?

damienmarchal commented:
According to the CI there is one test in Flexible that is failing in this PR.

matthieu-nesme commented:
Indeed and it is related to the size of a mstate
"Size of output dofs is wrong: 4 expected: 1"

matthieu-nesme commented:
I presume it has nothing to do with Flexible, but it rather seems related to the mstate size.

Maybe the test could be fixed, but what is important to notice is that this PR can have unexpected behaviors with existing code.

damienmarchal commented:
I really would like to have this merged. 
So how could we progress ? 
- can we remove the feature that makes the tests fail ? 
- do we fix the test and hope others person code will no be broken silently ? 
- can we implement kind of backward compatibility tricks ?

matthieu-nesme commented:
No way!

Did you have a chance to check what is wrong? It can be nothing as it can be fundamental, it has to be checked for sure.

damienmarchal commented:
Thanks for the clear answer, 

I'm afraid that if no one spend time on fixing this issue, the whole PR will wait for an un-defined amount of time. 
On my side I'm fully booked.

guparan commented:
I'm gonna find some time to investigate this test failure. To be continued...

damienmarchal commented:
A gentlereminder :)  @guparan

epernod committed:
Merge branch 'master' into issofa_topology

epernod commented:
I'll try, as soon as I can compile Sofa and all the plugins ( you know what I mean @hugtalbot  ;) )

epernod commented on SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl:
```diff
@@ -1157,10 +1160,12 @@ void MechanicalObject<DataTypes>::init()
     {
         // if a topology is present, implicitly copy position from it
-        if (m_topology != NULL && m_topology->getNbPoints() && m_topology->getContext() == this->getContext())
+        if (l_topology && l_topology->hasPos() /*&& l_topology->getContext() == this->getContext()*/ )
```
The test: ./Flexible_test --gtest_filter=TetrahedronVolumeMappingTest/0.test_perTetra
was failing because of this line and line 1090: 

The mapping_test is creating an input mechanicalObject with 4 dofs and an output mechanicalObject with 1 dof only. But only one MeshTopology is created in the parent node, with 4 points. 

Thus at the init, the output mechanicalObject was looking for a topology using: l_topology.set( this->getContext()->getActiveMeshTopology() ); 
which returns the parent topology with 4 points instead of 1.

So is it normal that this line doesn't check anymore the context? To fix the test, I set in the output mechanicalObject d_useTopology.setValue(false);
@JeremieA or @fjourdes could you confirm it is the right way of fixing this issue.

fjourdes replied:
Looking back at the code, I am wondering why the implementation on the trunk does not return a valid pointer to the MeshTopology. Ie why both the attempts of lines 1090 and 1093 (on the master branch) fail to recover it ? My point is that the crash should exist on the trunk as well. The default implementation of the ```get``` method also searches in the parent nodes.

fjourdes replied:
For me the only difference comes from the removal of  ```m_topology->getContext() == this->getContext()``` test (?), which initialise the position of the MechanicalObject **only** if the topology container reside in the local node.

epernod replied:
asking 3 times, won't make me work faster ;)

I think that in trunk and in this PR the line 1090/1093 return pointer to the MeshTopology of the parent Node (input topology of the mapping).
What change is indeed that the context of the MeshTopology is not checked anymore in line 1163. Uncommenting that test restore the trunk behavior. I.e the mapping_test pass into the else.

I'll check the d_volumePerNodes

fjourdes replied:
Intersetingly I could remove the triple comment, but not my wrong edit review text, which was supposed to stay as a comment. 
Do not know either why it was inserted 3 times.
Just as a reminder since now they all got removed.

Could you check that if in `TetrahedronVolumeMapping::apply` the output vector is resized properly ( ie by taking into account the value of `d_volumePerNodes` data, much like what is done during `init`) the crash is removed ?

epernod replied:
Ok I checked the `TetrahedronVolumeMapping::apply` value of `dOut`
In the test if `d_useTopology.setValue(false);`, size is 1. if not, test failed and size is 4. 
Same if I uncomment `m_topology->getContext() == this->getContext()` size is 1.

So verdict?

matthieu-nesme replied:
I do not understand why a mstate should be built from a parent topology?
It is luck that only TetrahedronVolumeMapping_test complains, but almost every situations should crash.
If you want to clean something, remove the implicit link to topology and force to explicitly give the topology to be init with: `<MObject topology="@../../topo"/>`, or even better remove topology from mstate and explicit everything: `<MObject position="@topo.position"/>` but this would break a lot of scenes.

damienmarchal replied:
I agree with matthieu that we should prefer promote the use of explicit link :)

epernod replied:
ok but do you want to do that in that PR or do we open another issue? 
If the test is restored, the behavior is similar to the trunk and we can merge the others fixs/modifs.

fjourdes reviewed (commented):
DONT KNOW HOWTO REMOVE WRONG EDIT
Could you check that if in `TetrahedronVolumeMapping::apply` the output vector is resized properly ( ie by taking into account the value of `d_volumePerNodes` data, much like what is done during `init`) the crash is removed ? 
DONT KNOW HOWTO REMOVE WRONG EDIT

JeremieA committed:
FIX: MappingTest now put the mapping in the child node along with its output state, rather than the root node.

This is required for several mechanisms to work correctly in Sofa, and was the root cause of the failure of the TetrahedronVolumeMapping test with the topology link change in MechanicalObject.

JeremieA commented:
Hello,

Before I explain the bug, let me give a bit of context (it is a bit long, but I hope it is useful ;)).

The point of this change was to make it transparent which topology is used by which state, and allow for correct initialization and most importantly topological changes propagation. One issue with the initial implementation, associating a state to the first topology found in its parents and then ignoring it if it is not in the same context (hopefully in all codes using state->m_topology, but who checked?), is that this make it impossible to share the same topology (and topological changes) with multiple states in different (children) nodes. One such example is when you use an IdentityMapping (or a Rigid <-> Vec3 mapping or any other mapping that produce the same number of outputs as its inputs), previously to get correct topology association and changes propagation you would need to replicate the topology (and the topological mappings) with a topology both before and after the mapping (but beware if they end up with different sizes, or updated in the wrong order...).

With our changes, there is a clear(er) method BaseContext::getActiveMeshTopology() that you can call from any context and will give you the topology that is relevant for this context, i.e. either the topology at this exact node, or from a parent node but only if there is no mapping in between (or if there are only mappings whose sameTopology() method returns true). Using this method, state->l_topology is initialized (and you can inspect its result in the GUI), and if it is not NULL it will be used by all computations for the state without extra hidden checks.

Now back to this bug, the last issue here was that the tested mapping (TetrahedronVolumeMapping) was not preserving the topology, but its output state was somehow linked to its input topology. The solution is not to go back to the old (broken) behavior, but instead to understand what is the actual root cause here. Theoretically, if the logic described above was behaving correctly, the output state would not be able to see the input topology because of the use of getActiveMeshTopology() and the fact that the mapping sameTopology() method would return false (thanks to its "safe" default implementation). The reason why this incorrect link was being created was because the mapping was created in the wrong Node. Normally, Sofa requires mappings to be in the same node as its output state. This is important in order to know for example which mapping is associated to which state (at least prior to the introduction of reflective Links, and still required because Mapping->State links are only one-way). This is used for example by visitors to know if a state is to be considered as real DOFs (because of the absence of a mapping in its node). But the MappingTest initialization code was not respecting this requirement, instead putting the mapping in the root node along with the input state (and the topology). Only the output state was in the child node. In this case, getActiveMeshTopology() from the output state does not see any mapping in the child, allowing to go to the parent to find a topology. It assumes that the Mapping in the root node is there to write to the state/topology of the root node, so it is not considered as something that would block the topology to remain active. Hence the wrong link being created, and the bug. The fix (beaaaf5) is a simple one-line change in MappingTest, creating the mapping in childNode instead of root, which should have been the case to respect Sofa's conventions (but which are unfortunately not checked/asserted by anyone).

Hopefully all is well now. It's difficult to know for sure because not-disabled tests are still failing on my computer (but they are also failing on the master...)

epernod committed:
Merge branch 'master' into issofa_topology

epernod committed:
Merge branch 'master' into issofa_topology

epernod committed:
[SofaKernel] FIX: bug in mechanicalObj using vsize instead of d_size

guparan commented:
[ci-build][with-scene-tests]

damienmarchal commented:
[ci-build]
I was passing by, noticed a PR ready...clicked on the merge button :)

JeremieA commented:
cool, but it looks like the merge button did not do its job :/

damienmarchal commented:
Oups I wanted  to make a rebuild against a recent master to be sure and forgot. Now done. :)

GG to everyone that make the PR & that helps in having it reviewed & merge
If someone notice something, please report asap on the issues board.
